Feature/elasticache user group association

ido-re opened the pull request:
<!--
Please read through https://git.io/fj2m9 if this is your first time opening a
pull request to this repo. Find us in https://crossplane.slack.com
if you need any help contributing.
-->

### Description of your changes

Allow Elasticach users to associate with Elasticach Usergroup.
This resource will allow you to create Elasticach (Redis) User and associate with an existing user group 

Important note: users with existing UserGroup resources must unset spec.forProvider.userIds before using the new UserGroupAssociation resource.

Fixes #1620

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Suggested test cases:

- [ ] A successful uptest run on the example manifest, which should leave `user_ids` unset on the user_group resource. Looking at the job output, the manifest during the input step should contain the user id from the association resource in `status.atProvider` but not in `spec.forProvider`.
- [ ] ...

[contribution process]: https://git.io/fj2m9

ido-re commented:
/test-examples="examples/elasticache/v1beta1/usergroupassociation.yaml"

ido-re commented:
/test-examples="examples/elasticache/v1beta1/usergroupassociation.yaml"

ido-re commented:
@ulucinar  @sergenyalcin  @turkenf  @mbbush - I think the PR is read for review :)

turkenf commented:
/test-examples="examples/elasticache/v1beta1/usergroupassociation.yaml"

ido-re commented:
@turkenf I see the uptest has failed, do I need to do something?

turkenf commented:
Hi @ido-re,
Congratulations on your first contribution 🎉 

> @turkenf I see the uptest has failed, do I need to do something?

You can check why it failed in the uptest [logs](https://github.com/crossplane-contrib/provider-upjet-aws/actions/runs/12950494669/job/36123417771). The test failed with the error below, sometimes the connection error can be temporary. I will trigger the test again. If it fails with the same error, please try using the v1beta1 versions of all resources in the example.
```
case.go:363: conversion webhook for elasticache.aws.upbound.io/v1beta2, Kind=User failed: Post "https://provider-aws-elasticache.upbound-system.svc:9443/convert?timeout=30s": dial tcp 10.96.250.111:9443: connect: connection refused
    case.go:363: conversion webhook for elasticache.aws.upbound.io/v1beta2, Kind=User failed: Post "https://provider-aws-elasticache.upbound-system.svc:9443/convert?timeout=30s": dial tcp 10.96.250.111:9443: connect: connection refused
```

turkenf commented:
/test-examples="examples/elasticache/v1beta1/usergroupassociation.yaml"

ido-re commented:
@turkenf - thank you, Following your suggestion, I changed the example files to use v1beta1

turkenf commented:
> @turkenf - thank you, Following your suggestion, I changed the example files to use v1beta1.

Thanks for the update @ido-re,

I think you have accidentally updated the generated example [here](https://github.com/crossplane-contrib/provider-upjet-aws/pull/1648/commits/09b69d92ce0f3270f4fd70c9c0c238c49dea752f). This is not required since we are not using the generated examples when testing.

Can you please change the default [User](https://github.com/crossplane-contrib/provider-upjet-aws/blob/09b69d92ce0f3270f4fd70c9c0c238c49dea752f/examples/elasticache/v1beta1/usergroupassociation.yaml#L21) resource version under the examples folder instead?

ido-re commented:
> > @turkenf - thank you, Following your suggestion, I changed the example files to use v1beta1.
> 
> Thanks for the update @ido-re,
> 
> I think you have accidentally updated the generated example [here](https://github.com/crossplane-contrib/provider-upjet-aws/pull/1648/commits/09b69d92ce0f3270f4fd70c9c0c238c49dea752f). This is not required since we are not using the generated examples when testing.
> 
> Can you please change the default [User](https://github.com/crossplane-contrib/provider-upjet-aws/blob/09b69d92ce0f3270f4fd70c9c0c238c49dea752f/examples/elasticache/v1beta1/usergroupassociation.yaml#L21) resource version under the examples folder instead?

ho I missed it, fixed it now @turkenf

ido-re commented:
> /test-examples="examples/elasticache/v1beta1/usergroupassociation.yaml"

@turkenf is it possible to run again the automation test? 
I rollbacked the changes I made on examples-generated/elasticache/v1beta1/usergroupassociation.yaml following your comment and because the diff test was failed

turkenf commented:
/test-examples="examples/elasticache/v1beta1/usergroupassociation.yaml"

turkenf commented:
> @turkenf I see the last failer is:
uptest-v0.13.0: error: cannot run e2e tests successfully: cannot execute tests: kuttl failed: exit status 1
Do you have suggestions for solving the issue?

If you scroll through the logs in the Uptest run, you can see messages in the conditions section of the resources about why the test failed. For example, I see the following message in the `UserGroupAssociation` conditions:

```
    logger.go:42: 10:04:35 | case/0-apply |     - lastTransitionTime: "2025-01-28T10:04:03Z"
    logger.go:42: 10:04:35 | case/0-apply |       message: 'create failed: async create failed: failed to create the resource:
    logger.go:42: 10:04:35 | case/0-apply |         [{0 creating ElastiCache User Group Association (example,example): operation
    logger.go:42: 10:04:35 | case/0-apply |         error ElastiCache: ModifyUserGroup, https response error StatusCode: 404,
    logger.go:42: 10:04:35 | case/0-apply |         RequestID: 9c32e451-ecc9-45bb-be9c-fb43aaf1b603, UserNotFound: User example
    logger.go:42: 10:04:35 | case/0-apply |         not found.  []}]'
```

My advice to you here is to first test/validate the resource manually. Please see:  [Running Uptest locally](https://github.com/crossplane/upjet/blob/main/docs/adding-new-resource.md#running-uptest-locally), [Manual Test](https://github.com/crossplane/upjet/blob/main/docs/adding-new-resource.md#manual-test)

ido-re commented:
> 2025-01-28T10:04:03Z

@turkenf - sorry for the late response - I tested it locally and I think the issue was found, can you please rerun the uptest?

jeanduplessis commented:
/test-examples="examples/elasticache/v1beta1/usergroupassociation.yaml"

ido-re commented:
@turkenf @jeanduplessis - I see failers of different tests even I rebase my commits, I'll be happy to get your direction

ido-re commented:
> @turkenf @jeanduplessis - I see failers of different tests even I rebase my commits, I'll be happy to get your direction

okay - I rebased again - @turkenf @jeanduplessis is it possible to rerun the tests?

jeanduplessis commented:
/test-examples="examples/elasticache/v1beta1/usergroupassociation.yaml"

ido-re commented:
@turkenf @turkenf  any chance to take a look? I have no idea why the check-diff is failed

turkenf commented:
@ido-re, for the check-diff error, please make sure you run `make generate` and push the changes.

ido-re commented:
> @ido-re, for the check-diff error, please make sure you run `make generate` and push the changes.

Yes - I did it, thank

ido-re commented:
> @ido-re, for the check-diff error, please make sure you run `make generate` and push the changes.

@turkenf can you please explain the steps? because I ran "make generate" before my last commit and still get the check-diff error
![image](https://github.com/user-attachments/assets/3cdb609d-fc4e-4062-abb3-ee8ca11a2e62)

turkenf commented:
@ido-re, please ignore the check-diff error for now; once this [PR](https://github.com/crossplane-contrib/provider-upjet-aws/pull/1674) is merged, you can rebase your branch, and the error will go away. I will trigger a test with a comment below, please do not make any changes until the uptest is completed to see the results.

turkenf commented:
/test-examples="examples/elasticache/v1beta1/usergroupassociation.yaml"

turkenf commented:
> @turkenf - sorry for the late response - I tested it locally and I think the issue was found, can you please rerun the uptest?

If you have successfully tested the resource manually, it would be very helpful if you could fill in the `How has this code been tested` section. Thank you 🙏

ido-re commented:
> > @turkenf - sorry for the late response - I tested it locally and I think the issue was found, can you please rerun the uptest?
> 
> If you have successfully tested the resource manually, it would be very helpful if you could fill in the `How has this code been tested` section. Thank you 🙏

@turkenf Because the failed is for the user creating, I tested it, and something so strange:
1. From some reason the uptest is running the API on ver v1beta2 when in the example I'm using v1beta1
ref from uptest:
` logger.go:42: 19:54:42 | case/0-apply | - apiVersion: elasticache.aws.upbound.io/v1beta2
    logger.go:42: 19:54:42 | case/0-apply |   kind: User
    logger.go:42: 19:54:42 | case/0-apply |   metadata:
    logger.go:42: 19:54:42 | case/0-apply |     annotations:
    logger.go:42: 19:54:42 | case/0-apply |       crossplane.io/external-create-failed: "2025-02-06T19:54:30Z"
'

2. In my environment I'm successful in creating the Redis user with the IAM type but here it failed 

ref from uptest:
'
    logger.go:42: 19:54:42 | case/0-apply |     - lastTransitionTime: "2025-02-06T19:41:30Z"
    logger.go:42: 19:54:42 | case/0-apply |       reason: Creating
    logger.go:42: 19:54:42 | case/0-apply |       status: "False"
    logger.go:42: 19:54:42 | case/0-apply |       type: Ready
    logger.go:42: 19:54:42 | case/0-apply |     - lastTransitionTime: "2025-02-06T19:54:30Z"
    logger.go:42: 19:54:42 | case/0-apply |       message: 'create failed: async create failed: failed to create the resource:
    logger.go:42: 19:54:42 | case/0-apply |         [{0 creating ElastiCache User (example): operation error ElastiCache: CreateUser,
    logger.go:42: 19:54:42 | case/0-apply |         https response error StatusCode: 400, RequestID: 37481c17-2f16-4e78-82f2-ee683d17f1f2,
    logger.go:42: 19:54:42 | case/0-apply |         InvalidParameterCombination: User Id and User name must be same for authentication
    logger.go:42: 19:54:42 | case/0-apply |         type: iam  []}]'
`

ref from my env:
'
apiVersion: elasticache.aws.upbound.io/v1beta1
kind: User
metadata:
  annotations:
    crossplane.io/external-name: authorization-euc1
  labels:
    argocd.argoproj.io/instance: authorization-stg-euc1-authorization
  name: elasticache-user-authorization
spec:
  forProvider:
    accessString: on ~* +@all
    authenticationMode:
      - type: iam
    engine: REDIS
    region: eu-central-1
    tags:
      managed-by: crossplane
    userName: authorization-euc1
'

turkenf commented:
Ok @ido-re, thanks for the explanation. When I find time, I will check your PR in detail and test it manually.

ido-re commented:
@turkenf finally I finished successfully running the local uptest (was blocked by the security team before) and now it's working properly with the results below:

-  new "test" user-group" created with existing "default" user
- new "exampleuser" created
- "exampleuser" attached to the "test" group and the "test" group includes 2 users "default" and "exampleuser"
- all the new resource terminated in the end of the uptest
- * test on my dev was account
the successfully example testes are ready for your remote uptest

turkenf commented:
/test-examples="examples/elasticache/v1beta1/usergroupassociation.yaml"

ido-re commented:
@turkenf I see all the tests are passed, should I need to do something?

turkenf commented:
@ido-re, I appreciate your effort here; no for now, thanks 🙏  I will review the changes in the coming days and do my best to add this resource to the minor release we will make at the end of February.

turkenf commented:
Hello @ido-re,

Before I started reviewing this PR, I noticed the following note in the Terraform documentation, which was a bit annoying:

<img width="668" alt="Screenshot 2025-02-12 at 17 05 10" src="https://github.com/user-attachments/assets/7a7e7281-d708-4dc4-92b2-aced84d9f24c" />

According to this note, it seems that the `aws_elasticache_user_group_association` resource may not work properly with the provider in its current state (to be confirmed with the team). It’s a bit frustrating to think that all the effort put into this PR might go to waste, but I appreciate the work done here.

Let me roughly explain the situation; since Crossplane continuously reconciles resources, there is a risk that `UserGroup` and `UserGroupAssociation` may conflict with each other. Specifically, `UserGroupAssociation` modifies the `userIds` of `UserGroup`, which Crossplane might then detect as a drift and attempt to revert, which can lead to an endless reconcile loop. I will share with you the results of the manual tests I conducted with the image I produced from this PR, which confirms this situation:

I created the resources `User`, `UserGroup` and `UserGroupAssociation`. The resources were created successfully and I verified in the console that users were added to the `UserGroup` resource:

![Screenshot 2025-02-12 at 16 46 01](https://github.com/user-attachments/assets/5eeca2d1-8ef1-4338-a2b2-f4dbb940c635)

After the first reconciliation, encountered the following diff in the logs and the `UserGroupAssociation` resources were set to `Ready=False`

```
2025-02-12T16:46:56+03:00	DEBUG	provider-aws	Diff detected	{"uid": "7840d46e-f1ba-4945-9559-036328179136", "name": "test", "gvk": "elasticache.aws.upbound.io/v1beta1, Kind=UserGroup", "instanceDiff": "*terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{\"user_ids.#\":*terraform.ResourceAttrDiff{Old:\"3\", New:\"1\", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, \"user_ids.1971754988\":*terraform.ResourceAttrDiff{Old:\"default\", New:\"default\", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, \"user_ids.236312002\":*terraform.ResourceAttrDiff{Old:\"exampleuser\", New:\"\", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, \"user_ids.3350137769\":*terraform.ResourceAttrDiff{Old:\"exampleuser2\", New:\"\", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyDeposed:false, DestroyTainted:false, RawConfig:cty.NilVal, RawState:cty.NilVal, RawPlan:cty.NilVal, Meta:map[string]interface {}(nil)}"}
```

And the ss from the console:

![Screenshot 2025-02-12 at 16 48 01](https://github.com/user-attachments/assets/729fb416-652e-4c0b-9114-db7bd65bcd1c)

After the next reconciliation, the `UserGroupAssociation` resource transitioned back to `Ready=True`, which indicates that it re-applies the association. However, since this cycle repeats on every reconcile, it effectively creates a race condition between `UserGroup` and `UserGroupAssociation`, preventing stable resource management. This behavior makes it difficult to use `UserGroupAssociation` reliably within Crossplane.

As I mentioned above, I will discuss this with the team to clarify the situation and update you accordingly. Thanks!

ido-re commented:
> Hello @ido-re,
> 
> Before I started reviewing this PR, I noticed the following note in the Terraform documentation, which was a bit annoying:
> 
> <img alt="Screenshot 2025-02-12 at 17 05 10" width="668" src="https://private-user-images.githubusercontent.com/103541666/412455547-7a7e7281-d708-4dc4-92b2-aced84d9f24c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk4Njg3MjQsIm5iZiI6MTczOTg2ODQyNCwicGF0aCI6Ii8xMDM1NDE2NjYvNDEyNDU1NTQ3LTdhN2U3MjgxLWQ3MDgtNGRjNC05MmIyLWFjZWQ4NGQ5ZjI0Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxOFQwODQ3MDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00Y2Q1NTM1ZmNmMzM5YWU3OGU3ODRjNzM2ZjAzNjQxZmZlM2VhMWNhOGUyMTdlMzljZjRkOGY3Nzc3NmVmNTNjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7bAlbW1a4q2RcN7YZuLgSS3BoiLu59I2_cBwh4gdIfU">
> According to this note, it seems that the `aws_elasticache_user_group_association` resource may not work properly with the provider in its current state (to be confirmed with the team). It’s a bit frustrating to think that all the effort put into this PR might go to waste, but I appreciate the work done here.
> 
> Let me roughly explain the situation; since Crossplane continuously reconciles resources, there is a risk that `UserGroup` and `UserGroupAssociation` may conflict with each other. Specifically, `UserGroupAssociation` modifies the `userIds` of `UserGroup`, which Crossplane might then detect as a drift and attempt to revert, which can lead to an endless reconcile loop. I will share with you the results of the manual tests I conducted with the image I produced from this PR, which confirms this situation:
> 
> I created the resources `User`, `UserGroup` and `UserGroupAssociation`. The resources were created successfully and I verified in the console that users were added to the `UserGroup` resource:
> 
> ![Screenshot 2025-02-12 at 16 46 01](https://private-user-images.githubusercontent.com/103541666/412464833-5eeca2d1-8ef1-4338-a2b2-f4dbb940c635.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk4Njg3MjQsIm5iZiI6MTczOTg2ODQyNCwicGF0aCI6Ii8xMDM1NDE2NjYvNDEyNDY0ODMzLTVlZWNhMmQxLThlZjEtNDMzOC1hMmIyLWY0ZGJiOTQwYzYzNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxOFQwODQ3MDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xZDAwYjdlZTdlZjhlNDQ0ZjBkYTk0YzFjNTA3MzQyMDAxZDk5YjQ0MWNhNGNiNzM5MjNhMjEwOTNjZjhlM2NjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.hXkTd2QqOcA8Eaw6w6IevpAxRCaFF1f25v-BEAbBxpY)
> 
> After the first reconciliation, encountered the following diff in the logs and the `UserGroupAssociation` resources were set to `Ready=False`
> 
> ```
> 2025-02-12T16:46:56+03:00	DEBUG	provider-aws	Diff detected	{"uid": "7840d46e-f1ba-4945-9559-036328179136", "name": "test", "gvk": "elasticache.aws.upbound.io/v1beta1, Kind=UserGroup", "instanceDiff": "*terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{\"user_ids.#\":*terraform.ResourceAttrDiff{Old:\"3\", New:\"1\", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, \"user_ids.1971754988\":*terraform.ResourceAttrDiff{Old:\"default\", New:\"default\", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, \"user_ids.236312002\":*terraform.ResourceAttrDiff{Old:\"exampleuser\", New:\"\", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, \"user_ids.3350137769\":*terraform.ResourceAttrDiff{Old:\"exampleuser2\", New:\"\", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyDeposed:false, DestroyTainted:false, RawConfig:cty.NilVal, RawState:cty.NilVal, RawPlan:cty.NilVal, Meta:map[string]interface {}(nil)}"}
> ```
> 
> And the ss from the console:
> 
> ![Screenshot 2025-02-12 at 16 48 01](https://private-user-images.githubusercontent.com/103541666/412466168-729fb416-652e-4c0b-9114-db7bd65bcd1c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk4Njg3MjQsIm5iZiI6MTczOTg2ODQyNCwicGF0aCI6Ii8xMDM1NDE2NjYvNDEyNDY2MTY4LTcyOWZiNDE2LTY1MmUtNGMwYi05MTE0LWRiN2JkNjViY2QxYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxOFQwODQ3MDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYjJjMDk5NjI5NjFhNjgzODc0MzA2MjE1MWM5ZTQ5MjBhNWUxNmIwZDE3MzM5MDNjMTA5Y2I2YzI5NmQxZmI3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.72iMh42lfr5pm7aj38lkzA3Wwd-Q-a05vtlNy-kx6rs)
> 
> After the next reconciliation, the `UserGroupAssociation` resource transitioned back to `Ready=True`, which indicates that it re-applies the association. However, since this cycle repeats on every reconcile, it effectively creates a race condition between `UserGroup` and `UserGroupAssociation`, preventing stable resource management. This behavior makes it difficult to use `UserGroupAssociation` reliably within Crossplane.
> 
> As I mentioned above, I will discuss this with the team to clarify the situation and update you accordingly. Thanks!

@turkenf is the "initProvider" can help in this case?

turkenf commented:
> @turkenf is the "initProvider" section can help in this case?

@ido-re, using `initProvider` can be logical. Could you please test it and share the results with us? If it works, it would be great if you could add a note to the example YAML and update it.

ido-re commented:
> > @turkenf is the "initProvider" section can help in this case?
> 
> @ido-re, using `initProvider` can be logical. Could you please test it and share the results with us? If it works, it would be great if you could add a note to the example YAML and update it.

Added one more user to check the ignore changes + use initProvider on the UserGroup
Also test it locally via the e2e tests.

@turkenf - can you please run the remote tests?

mbbush commented on examples/elasticache/v1beta1/usergroupassociation.yaml:
```diff
@@ -0,0 +1,98 @@
+  forProvider:
+    engine: REDIS
+    region: us-west-1
+  initProvider: # If using "UserGroupAssociation" -  use initProvider to create the user group but ignore changes
```
This feels like we're setting a trap for unsuspecting users that don't see this comment. I'm not sure if this will even show up in the example manifests in the registry. 

If the issue is a conflict between `UserGroup.spec.forProvider.userIds` and `UserGroupAssociation`, why don't we instead just configure the `UserGroup` resource to skip LateInit of `userIds`, and add a sentence to the description of that field describing why?

mbbush replied:
[Example of adjusting the generated description](https://github.com/crossplane-contrib/provider-upjet-aws/blob/45ec02187ac6371229b719fc3eb277914a666552/config/kafka/config.go#L234)

[Example of skipping LateInit for certain fields](https://github.com/crossplane-contrib/provider-upjet-aws/blob/45ec02187ac6371229b719fc3eb277914a666552/config/iam/config.go#L39)

ido-re replied:
@mbbush, Thank you for your inputs.
I added the init provider following @turkenf attention, about Terrafrom documentation note [HERE](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/elasticache_user_group_association)

The documentation provides information about a conflict when both the "**user-group**" and "**user-group-attachment**" are managed by IaC.

Ignoring users when one of the resources is not managed by Crossplane will prevent user management by IaC.
Example: If all users are managed via the "**user-group**" resource, I can't ignore it.

Additionally, in some cases, the "**user-group**" is managed by a different IaC system. Consider the following scenario:

Terraform – when creating the "**user-group**" as part of the ElastiCache cluster creation.
Crossplane – when users are attached to the "**user-group**" from different states. For example, Crossplane may attach users when a new K8s service is created with IAM and a service account.
I’d appreciate your input on this.

Regarding the comment, could you please clarify how I can add the description to the new resource documentation?

Thanks :)

mbbush replied:
This boils down to the principle that there should be exactly one way to manage each resource. At the moment, the `UserGroup` resource manages the relationship between users and groups, in a single, hard-to-manage list with a selector.

As you've identified, it would be far better to have an individual resource for each user-group association, which is what you've created. Thank you for the contribution! I've had to work around the non-existence of this sort of resource before, and while I found a "clever" workaround, it was a counterintuitive and brittle hack. 

The trouble comes from needing to ensure the new, easy-to-use resource doesn't cause conflicts with the existing resource. I ran into [this exact issue](https://github.com/crossplane-contrib/provider-upjet-aws/issues/929) a long time back with IAM roles and role policy associations. 

The situation here is slightly different, as the user group association resource does not yet exist, which means "don't add the user group association resource" is a possible resolution, but I don't want to take this approach if we can find an acceptable alternative. 

Here's my suggestion:
* Update the UserGroup resource to not late initialize the user ids list. 
* Update the field documentation on the UserGroup.userIds and the resource documentation on UserGroupAssociation to indicate that the resource conflicts with the field, and UserGroup.spec.forProvider.userIds must be left null if you want to use the UserGroupAssociation resource. 
* Mention in the release notes that users with existing UserGroup resources will need to unset `spec.forProvider.userIds` before they can use the new UserGroupAssociation resource.

All of these changes (except the release notes) need to be done in `config/elasticache/config.go` in order for them to actually have effect and survive the codegen process. 

Is it all right with you if I put some suggested test cases in the PR description?

ido-re replied:
@mbbush – Thanks for replying and clarifying the issue.
I made the changes in `config/elasticache/config.go` and would appreciate your review.
I also updated the description (if the release note should be changed differently, please let me know).

Regarding the test suggestion – yeah, go ahead! 😊

mbbush commented:
/test-examples="examples/elasticache/v1beta1/usergroupassociation.yaml"

ido-re commented on package/crds/elasticache.aws.upbound.io_usergroupassociations.yaml:
```diff
@@ -0,0 +1,651 @@
+    name: v1beta1
+    schema:
+      openAPIV3Schema:
+        description: UserGroupAssociation is the Schema for the UserGroupAssociations
```
@turkenf @mbbush - is that okay?
```suggestion
         description: |-
            UserGroupAssociation is the Schema for the UserGroupAssociations API. Associate an ElastiCache user and user group.
            NOTE:
            The API will detect changes in the UserGroups since UserGroupAssociations changes the user IDs associated with the user group.
            You can ignore these changes with the initProvider section as shown in the example.
```

mbbush replied:
You can't edit these files manually: they're generated. Instead you need to modify them in code while generation is happening. That's what https://github.com/crossplane-contrib/provider-upjet-aws/blob/45ec02187ac6371229b719fc3eb277914a666552/config/kafka/config.go#L234 does. As far as what the docs should say, that will depend on how we resolve the conflict between the two resources.

ido-re commented:
@mbbush – Thanks for replying and clarifying the issue.
I made the changes in config/elasticache/config.go and would appreciate your review.
I also updated the description (if the release note should be changed differently, please let me know).

Regarding the test suggestion – yeah, go ahead! 😊

mbbush commented:
You will need to run `make generate` locally and commit the results. https://github.com/crossplane-contrib/provider-upjet-aws/actions/runs/13864711940/job/39132677619#step:7:55

The config changes look like they're on the right track.

mbbush commented:
The number of commits was a problem. I rebased and squashed everything down into two: one with the changes to handwritten files and one with the changes to the generated files.

mbbush commented on config/elasticache/config.go:
```diff
@@ -223,5 +223,17 @@ func Configure(p *config.Provider) { //nolint:gocyclo
+	})
+
+	p.AddResourceConfigurator("aws_elasticache_user_group_association", func(r *config.Resource) {
+		r.MetaResource.ArgumentDocs[""] = "Note: This resource conflicts with `aws_elasticache_user_group` when `user_ids` is provided in the `aws_elasticache_user_group` resource. To avoid conflict, `UserGroup.spec.forProvider.userIds` must be left null."
```
This line as written does nothing. It's just adding an element to the map with a key of empty string, which nothing will ever read. 

Try appending to `r.MetaResource.Description`. You'll know it worked when you see it show up in the generated CRD after running make generate.

If you want to know what things you can change, take a look at the yaml file in `config/provider-metadata.yaml`. That's a generated file, so you can't edit it there, but it will help you see what the schema is for `r.MetaResource` (except the casing convention is changed)

mbbush commented:
/test-examples="examples/elasticache/v1beta1/usergroupassociation.yaml"

mbbush commented:
@ido-re For future changes, you should feel free to force-push to your branch over my commits. Just make sure that you've added a "Signed-off-by" message at the end, which you get "for free" using the `-s` flag to `git commit`. 

Do you have a working local test environment set up? You should be able to see this work by running `make e2e UPTEST_EXAMPLE_LIST=examples/elasticache/v1beta1/usergroupassociation.yaml`. See the comments in the makefile about how to configure credentials: you need an env var set to a multi-line value.

ido-re commented:
> @ido-re For future changes, you should feel free to force-push to your branch over my commits. Just make sure that you've added a "Signed-off-by" message at the end, which you get "for free" using the `-s` flag to `git commit`.
> 
> Do you have a working local test environment set up? You should be able to see this work by running `make e2e UPTEST_EXAMPLE_LIST=examples/elasticache/v1beta1/usergroupassociation.yaml`. See the comments in the makefile about how to configure credentials: you need an env var set to a multi-line value.

@mbbush Sure,
About the Uptest, yes, here are the results:

`    logger.go:42: 08:49:46 | case | case events from ns kuttl-test-pro-louse:
    logger.go:42: 08:49:46 | case | Deleting namespace: kuttl-test-pro-louse
=== CONT  kuttl
    harness.go:402: run tests finished
    harness.go:511: cleaning up
    harness.go:553: skipping cluster tear down
    harness.go:554: to connect to the cluster, run: export KUBECONFIG="/Users/ido.r/Documents/git_personal/provider-upjet-aws/kubeconfig"
--- PASS: kuttl (716.43s)
    --- PASS: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/case (715.79s)
PASS
08:49:46 [ OK ] running automated tests`

![image](https://github.com/user-attachments/assets/bdaa3b3a-22b0-420e-bd9f-93b47babcae9)

ido-re commented:
@mbbush @turkenf - I'll be happy to get your attention :)

mbbush commented:
@ido-re I don't see any new commits since https://github.com/crossplane-contrib/provider-upjet-aws/pull/1648#discussion_r2011047468

ido-re commented:
> @ido-re I don't see any new commits since [#1648 (comment)](https://github.com/crossplane-contrib/provider-upjet-aws/pull/1648#discussion_r2011047468)

I see. It is very strange because my last commit included rebase, which creates about 49 commits.
Anyway, this PR includes the changes you asked me to make.
The changes were also tested via the e2e test locally,

ido-re commented:
@mbbush - I'll be happy for your replay following our discussion here

mbbush commented:
> @mbbush I see. It is very strange because my last commit included rebase, which creates about 49 commits. It may be related to your commit, which was the last one?
Anyway, this PR includes the changes you asked me to make.
The changes were also tested via the e2e test locally (in answer to your last question),

The only commits on the source branch of this PR are the two commits I created by squashing your changes to non-generated and generated files. The changes I requested in https://github.com/crossplane-contrib/provider-upjet-aws/pull/1648#discussion_r2011047468 are not present. 

Maybe you changed something locally and didn't push it up to your fork, or the push was rejected because it contained my commits?

It's fine for you to blow away my commits, I just want a reasonable number of commits to review, so if you do, please squash your commits together.

ido-re commented:
@mbbush Is it all right with you if I create a new PR from the head with the changes you asked for?
